### ISSUE TITLE
allow fetching of lazy relations outside of its original transaction

### ISSUE THREAD

**Issue by ymajoros** [quarkusio/quarkus#38916]
### Description

As a follow-up to https://github.com/quarkusio/quarkus/issues/38807, I think it would be very useful to be able to access lazy relations of detached entities:

- Spring has "Open session in view".
- Wildfly seems to have a similar mechanism by default (our application is migrating from Wildfly and the problem wasn't present).
- Eclipselink (JPA RI) also allows to access lazy properties of detached entities by default (not even sure you can disable that).
- enable_lazy_load_no_trans is mentioned in multiple sources, although I tried to activate it with quarkus.hibernate-orm.unsupported-properties."hibernate.enable_lazy_load_no_trans"=true, which is actually being read, but I couldn't find any usage besides configuration in Hibernate 6.4.
- [Some](https://vladmihalcea.com/the-open-session-in-view-anti-pattern/) consider it an antipattern, [others](https://www.baeldung.com/spring-open-session-in-view#2-pattern-or-anti-pattern) have a more nuanced perspective.

Example use case for which it would be practically needed: chained relations, e.g.  a `Statement` has a `previousStatement`; you want to be able to fetch the previous statement, you'll maybe need the next one afterwards etc. Performance is often mentioned in the CONs, but you would end up fetching the objects one by one when needed anyway, detached entity or not, and an open-in-view mechanism wouldn't make any difference there. On the contrary, it will in some situations be more performant: the relation will only be queried from the database if/when needed instead of being fetched in all cases, eagerly or from a projection.

I suggest implementing a similar feature in Quarkus.

**Comment by yrodiere** [comment]
Hey @beikov , could you please share your opinion on the right solution for this using Hibernate ORM? I'd definitely be interested in hearing it.

**Comment by beikov** [comment]
I think the proper solution is to always think about the what data you really need and model this with DTOs to avoid any sort of unpredictable lazy loading.
In this particular case, you seem to need some sort of recursive fetching, so I would suggest you to make use of recursive CTEs to model the data loading e.g.

```sql
with statements as (
  select cast(:id) as id
  union all
  select s.previousStatement.id as id from Statement s join statements r on s.id = r.id
)
select s from Statement s where s.id in (select r.id from statements r)
```

This way, you will select all statements that are reachable via `previousStatement` associations, starting at the statement with the id as given through the `:id` parameter.

Ideally, we'd offer a [recursive fetch join](https://hibernate.atlassian.net/browse/HHH-16187) to model this, but we don't have that yet.

**Comment by ymajoros** [comment]
@beikov thanks for your feedback. In my use case:
- I most of the time don't need the previousStatement
- ... but some flows do
- ... and sometimes we need to dig further

As getting a statement is a O(1) operation, but fetching the whole chain is O(n), I do not want to fetch them eagerly or with any kind of projection or recursive fetch. Fetching lazily means that I have a best case of O(1) and a worst case of O(f) (f meaning here a few). The proposed alternative is O(n) best case, making it perform much worse (n can be big here).

There are a lot of other valid use cases. Like the 'debate' about open-session-in-view for spring, there will always be PROs and CONs, and I suggest to not go too deep into this debate here but let users make conscious decisions about their use cases.

Therefore, such a functionality is useful for some users.

**Comment by beikov** [comment]
Then IMO you should model your different flows explicitly by creating separate data access methods that initializes exactly the state that the caller needs. Any sort of unpredictable lazy initialization will lead to pain, so it's best to avoid it in the first place.
Open-session-in-view or enabling the [`hibernate.enable_lazy_load_no_trans`](https://docs.jboss.org/hibernate/orm/6.4/userguide/html_single/Hibernate_User_Guide.html#settings-hibernate.enable_lazy_load_no_trans) setting are both recipes for disaster.

>but fetching the whole chain is O(n)

Talking about `O(n)` here is a bit confusing, because the expensive part really is the separate select statement execution, not the lookup in the database itself. Using a recursive fetch like I showed you only does one database request.

>There are a lot of other valid use cases. Like the 'debate' about open-session-in-view for spring, there will always be PROs and CONs, and I suggest to not go too deep into this debate here but let users make conscious decisions about their use cases.

The only PRO I see is that you can turn off your brain and not think about the data access design, but that will bite you very fast when you run into performance issues, because changing the design later is painful. So I don't really view that as a PRO, but rather as a CON in disguise. Modelling your data fetching requirements is IMO not something that should be an after-thought just like performance testing shouldn't be, because at some point there will be that one user with hundreds of chained `Statement` associations who will bring down your whole application.

>Therefore, such a functionality is useful for some users.

If you really want this broken behavior, then let the transaction span across the whole request by annotating your HTTP endpoint method with `@Transactional`. Nobody is preventing you from doing that. Having a global flag is IMO a very bad idea as that will make every HTTP request claim a database connection and also span a transaction including locks for much longer than needed.

**Comment by ymajoros** [comment]
This is quite opinionated, and I really have a different view and experience.

Even if the recursive fetch is a single request, it is definitely a O(n) one that I want to avoid. As said, I most often do not need this data.

Of course, we could replace the relation with an ad-hoc call. But that also means refactoring the database to replace the many-to-one by a relation table, which I don't consider better design for this use case.

'Turn off your brain' is rather though wording: I'd rather rephrase it as decoupling the service returning a Statement from whatever the next transaction wants to do with it.

There is no use case where a user will bring down the whole application, for multiple reasons.

This is also a very simple use case, but navigating complex object graphs not known in advance in a different transaction isn't bad design per se. Performance isn't an afterthought here, on the contrary.

I'm new to this project, but having bigger transactions is one of the steps that we are taking further. Still, it doesn't cover all use cases and we sometimes do want a separate one.

Of course, https://hibernate.atlassian.net/browse/HHH-17750 makes our problem worse: if fetching a lazy transient attribute of a detached entity would work in a new transaction, the need for this would be much lower (though still there).

**Comment by beikov** [comment]
>This is quite opinionated, and I really have a different view and experience.

I was asked for an opinion and here you have it. I simply prefer designing the data access rather than burying requirements in e.g. the UI code through implicit lazy initialization. If you ask me, no entity should ever be serialized to JSON, but only DTOs. I understand that this might seem extreme to some folks, but designing the data access this way makes it way simpler to reason about performance and optimize, simply because there are no unknowns.

>Even if the recursive fetch is a single request, it is definitely a O(n) one that I want to avoid. As said, I most often do not need this data.

That's totally understandable, hence my suggestion.

>Then IMO you should model your different flows explicitly by creating separate data access methods that initializes exactly the state that the caller needs.


>Of course, we could replace the relation with an ad-hoc call. But that also means refactoring the database to replace the many-to-one by a relation table, which I don't consider better design for this use case.

I didn't suggest that and I agree that this would not improve the design.

>'Turn off your brain' is rather though wording: I'd rather rephrase it as decoupling the service returning a Statement from whatever the next transaction wants to do with it.

I don't know your application design, but it seems to me that your app could know beforehand what the next transaction requires and hence load the appropriate data without the need for lazy loading. It's a matter of creating separate data access methods for the respective flows.

>This is also a very simple use case, but navigating complex object graphs not known in advance in a different transaction isn't bad design per se. Performance isn't an afterthought here, on the contrary.
>I'm new to this project, but having bigger transactions is one of the steps that we are taking further. Still, it doesn't cover all use cases and we sometimes do want a separate one.

No idea how you are "navigating" in "a different transaction", but maybe you you can solve your issues by simply loading the subsequent object graphs explicitly instead of relying on lazy loading?

>Of course, https://hibernate.atlassian.net/browse/HHH-17750 makes our problem worse: if fetching a lazy transient attribute of a detached entity would work in a new transaction, the need for this would be much lower (though still there).

It seems @mbladel is already looking into this issue, so let's maybe keep the discussion about the Jira issue in the Jira comments.

**Comment by ymajoros** [comment]
> I simply prefer designing the data access rather than burying requirements in e.g. the UI code through implicit lazy initialization. 

This isn't UI code. The first transaction is simply committed and we end up with an object with lazy relations.

> If you ask me, no entity should ever be serialized to JSON, but only DTOs. 

I do strictly agree with that. Not what is happening in my example.

> I understand that this might seem extreme to some folks, but designing the data access this way makes it way simpler to reason about performance and optimize, simply because there are no unknowns.
>

There are unknowns, the first service doesn't know how much data it has to send back. Even the caller doesn't know at call time, it might need to fetch relations. It's the essence of lazy relationships, the fact they are in a different transaction doesn't make this different.

> > Even if the recursive fetch is a single request, it is definitely a O(n) one that I want to avoid. As said, I most often do not need this data.
> 
> That's totally understandable, hence my suggestion.
> 
> > Then IMO you should model your different flows explicitly by creating separate data access methods that initializes exactly the state that the caller needs.
> 

No, at the time of the first call, I can't decide how many previous statements I need. It also depends on expensive calculations on the statement itself (calling third party services, ...). I really can't know beforehand.

> I don't know your application design, but it seems to me that your app could know beforehand what the next transaction requires and hence load the appropriate data without the need for lazy loading. It's a matter of creating separate data access methods for the respective flows.
> 

So, basically, this is not the case.

I also don't see different kind of structures for slightly different use cases as a better design. Your proposal does have its merits to optimize some use cases, but isn't the holy grail either.

JPA is a leaky abstraction, and that's both a problem and some of its strengths. Lazy loading, even in a different transaction, is quite useful in some scenarios. I do agree it can be misused.

> No idea how you are "navigating" in "a different transaction", but maybe you you can solve your issues by simply loading the subsequent object graphs explicitly instead of relying on lazy loading?

That means you need to know all the paths that you'll have to follow beforehand. If there is any complex logic or dependencies, this isn't necessarily the case.

> It seems @mbladel is already looking into this issue, so let's maybe keep the discussion about the Jira issue in the Jira comments.

Yes. I just meant this is the first reason why we wanted this feature, even it were only for migration.

**Comment by ymajoros** [comment]
> Maybe I'm not totally understanding, but in your second transaction, you can execute your logic to determine what data needs to be fetched and then load the `Statement` with the id of the `previousStatement` along with how many `previousStatement` you may need, but in one go. Put all of that into a DTO so there is no lazy entity anymore and you're done :)

In the second transaction, I can't know if I need a 3rd or 4th object before checking the second one. How would an ad-hoc call to fetch a new statement be different from a lazy fetch regarding performance? I'm still in the service layer btw.

**Comment by ymajoros** [comment]
To be honest: though different point of views on detached entities are somewhat interesting, this is irrelevant here as this is what we choose to use in our own projects. Being able to merge detached entities with lazy fields sounds like a basic use case, regardless of how other people like the pattern or not. JPA is a leaky abstraction, and the arguments I heard here seem to be against some core functionalities of JPA. This doesn't seem like the best place for this discussion IMHO.

Anyway, it seems this was considered as a Hibernate bug and fixed in the upcoming 6.4.5 release, which looks almost ready in Jira: see https://hibernate.atlassian.net/browse/HHH-17750 & https://hibernate.atlassian.net/projects/HHH/versions/32262/tab/release-report-all-issues

So, I guess the next step is now to simply wait on that release + integration of Hibernate core 6.4.5 in Quarkus?

**Comment by yrodiere** [comment]
> So, I guess the next step is now to simply wait on that release + integration of Hibernate core 6.4.5 in Quarkus?

Correct. That will lead to closing #38807, while this specific issue (open session in view for Quarkus, essentially) will still be open for discussion / up for grab.

**Comment by PizzaApps** [comment]
We are in the same situation as @lazyBisa and have been blocked in our migration to Quarkus because the application made heavy use of lazy loading with eclipselink.

As i found a way to make it work with Quarkus 3.24.1, I wanted to share my solution as it may be helpful for others in the same situation.

It consists of two parts:
1. The property ``quarkus.hibernate-orm.unsupported-properties."hibernate.enable_lazy_load_no_trans"=true`` must be set
2. The SessionFactory used by Quarkus must be added to the SessionFactoryRegistry, otherwise temporary sessions cannot be constructed for lazy loading

Code snippet for the second part:
```java
@ApplicationScoped
@IfBuildProperty(name = "quarkus.hibernate-orm.unsupported-properties.\"hibernate.enable_lazy_load_no_trans\"", stringValue = "true")
public class SessionFactoryRegistrator {

  @Inject
  SessionFactory sessionFactory;

  void onStart(@Observes StartupEvent ev) {
    Log.info("Adding Quarkus session factory to Hibernate session factory registry");

    SessionFactory unwrappedFactory = this.sessionFactory;
    if (ClientProxy.class.isAssignableFrom(this.sessionFactory.getClass())) {
      unwrappedFactory = (SessionFactory) ((ClientProxy) this.sessionFactory).arc_contextualInstance();
    }

    var sessionFactoryImpl = (SessionFactoryImpl) unwrappedFactory;
    var jndiService = sessionFactoryImpl.getServiceRegistry()
        .getService(org.hibernate.engine.jndi.spi.JndiService.class);

    // Manually register the session factory
    SessionFactoryRegistry.INSTANCE.addSessionFactory(
        sessionFactoryImpl.getUuid(),
        this.sessionFactory.getName(),
        this.sessionFactory.getJndiName(),
        sessionFactoryImpl,
        jndiService
    );
  }

}
```

**Comment by beikov** [comment]
>I'm new to this project, but having bigger transactions is one of the steps that we are taking further. Still, it doesn't cover all use cases and we sometimes do want a separate one.

@ymajoros can you please explain why you might want separate transactions? I have a clue about what this might be, but I would genuinely like to hear your reasons to understand what might be a better way forward.

**Comment by ymajoros** [comment]
_I have a clue about what this might be, but I would genuinely like to hear your reasons to understand what might be a better way forward._

What do you mean?

This discussion is almost 2y old, so a lot has changed. Still, we do have a lot of uses for separate transactions and, back to the subject, fetching lazy relations outside of the original transaction is kind of a basic feature. This isn't an afterthought but some basic, valid design and use of lazy relations: avoiding fetching a big object graph in advance that you might not even need.

**Comment by beikov** [comment]
I'm not arguing that you should get rid of it. I want to understand what the exact use case is. Why don't you want to have one transaction spanning across all interactions?

**Comment by ymajoros** [comment]
We do have a single one in most flows. We don't in some others though. 🥲 

To give a very simple example, back to the very definition of a transaction: in some flows, we want multiple blocks of operations to succeed as a whole, even if the next block fails.

**Comment by yrodiere** [comment]
> To give a very simple example, back to the very definition of a transaction: in some flows, we want multiple blocks of operations to succeed as a whole, even if the next block fails.

In that case the recommendation would be to reattach entities to the new session in the new transaction before re-using them -- basically call `session.refresh()`. You don't want to use that because... it's impractical in your case? Many entities perhaps? Something else?

**Comment by beikov** [comment]
Well, in that case, an [EXTENDED scoped persistence context](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a2027) might be more appropriate for you. You could track https://github.com/quarkusio/quarkus/issues/32940 to get updates on when the feature becomes available.

>In that case the recommendation would be to reattach entities to the new session in the new transaction before re-using them -- basically call session.refresh(). You don't want to use that because... it's impractical in your case? Many entities perhaps? Something else?

Calling `refresh()` is only allowed on managed entities, so that's not an option. He'd have to reload all the entities by a query that are needed, since there is no way to "re-attach" entities other than `merge`, which might in fact return a new object graph.

**Comment by yrodiere** [comment]
> Calling `refresh()` is only allowed on managed entities, so that's not an option. He'd have to reload all the entities by a query that are needed, since there is no way to "re-attach" entities other than `merge`, which might in fact return a new object graph.

Ah, right, sorry. This actually brings back memories from when I had to do this in the past, and essentially I was just doing something like `entity = session.getReference(entity.getClass(), entity.getId())`.
Which can be fine in simple cases, or near impossible if references to entities are all over the place (e.g. nested in a complex data structure).
EDIT: And, of course, it will load the data from the DB again -- but arguably that's necessary if you're in a new transaction and want to be in sync with the DB.

**Comment by beikov** [comment]
>EDIT: And, of course, it will load the data from the DB again -- but arguably that's necessary if you're in a new transaction and want to be in sync with the DB.

I partly agree. If the transaction isolation is repeatable read or serializable, then it might make a difference, but you can still get stale data with read committed isolation, since a follow-up find will not refresh an entity. If you want to avoid concurrency related problems, you have to resort to pessimistic or optimistic locks.

**Comment by ymajoros** [comment]
> Well, in that case, an [EXTENDED scoped persistence context](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a2027) might be more appropriate for you. You could track [#32940](https://github.com/quarkusio/quarkus/issues/32940) to get updates on when the feature becomes available.

No, I absolutely don't want that. I want to have transaction-scoped entity managers with clear transaction boundaries and no manual house keeping.

I just want to be able to get a lazy relation whenever needed, and not before. That's just what _lazy_ means. It should work even if I'm in a new transaction, there is no reason for it not to. And it now does perfectly btw (it didn't at the time I created this ticket).

I think we should close this ticket: 
- The bugs have long be fixed and ported to quarkus.
- The discussion evolved to "I don't like @Lazy outside of the current transaction and you shouldn't either" (sorry but this is how it feels). I heard your opinion, but my use case is still valid and trivial (and it works, now).

**Comment by beikov** [comment]
>No, I absolutely don't want that. I want to have transaction-scoped entity managers with clear transaction boundaries and no manual house keeping.

You don't have any "manual house keeping" to do and transaction boundaries are very clear. In Jakarta EE, you could scope an EXTENDED scoped `EntityManager` as e.g. `@RequestScoped` and have it work within multiple `@Transactional` methods, automatically joining the transaction and flushing at commit time.

>I just want to be able to get a lazy relation whenever needed, and not before. That's just what lazy means. It should work even if I'm in a new transaction, there is no reason for it not to. And it now does perfectly btw (it didn't at the time I created this ticket).

I understand that the problem with the `merge` you have originally reported is solved now, and that's good. I just wanted to understand your use case a bit better, because I think users migrating from EclipseLink to Hibernate ORM might have similar needs. I wanted to see what your motivation is for doing this, to understand if there is something we are missing.

>The discussion evolved to "I don't like @lazy outside of the current transaction and you shouldn't either" (sorry but this is how it feels). I heard your opinion, but my use case is still valid and trivial (and it works, now).

I'm not saying that the goal you want to achieve is invalid, just that the approach you're taking might not be the best fit. Given that an EXTENDED scoped persistence context would avoid all the reloading in separate transactions, but at the same time be nicely integrated with transactions, I feel like this would be a good solution for your use case as well as anyone else who wants to do lazy loading after a transaction ended.

I didn't want to put words in your mouth, which is why I tried to ask the open question about your use case. My one guess about your requirements was that you're using separate transactions to reduce the time a JDBC connection is held. And if reducing the connection hold time were something you wanted to achieve, I would have told you that we're thinking about a "read-only" mode that releases connections immediately.

Are you actually changing the database state in these individual transactions or are you just doing lazy loading? If you're doing just lazy loading, then this "read-only" mode might be a good alternative to an EXTENDED scoped persistence context to avoid holding a connection for longer than needed, though I think it may be possible to release connections immediately for EXTENDED scoped persistence contexts as well.
